File: src/main/java/phoenix/world/StageManager.java
package phoenix.world;

import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.block.PaneBlock;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.MathHelper;
import net.minecraft.world.IWorld;
import net.minecraft.world.World;
import phoenix.init.PhoenixBlocks;
import phoenix.world.structures.CustomEndSpike;

public class StageManager
{
    private static CompoundNBT data = new CompoundNBT();

    public static void read(CompoundNBT nbt)
    {
        if (nbt.contains("stage_nbt"))
        {
            data = nbt.getCompound("stage_nbt");
        }
        else
        {
            CompoundNBT compound = new CompoundNBT();
            compound.putInt("stage", 0);
            compound.putInt("part",  0);
            data = compound;
            nbt.put("stage_nbt", compound);
        }
    }

    public static CompoundNBT write(CompoundNBT compound)
    {
        compound.put("stage_nbt", data);
        return compound;
    }

    public static int getStage()
    {
        if(data == null) data = new CompoundNBT();
        return data.getInt("stage");
    }

    public static Stage getStageEnum()
    {
        if(data == null) data = new CompoundNBT();
        return Stage.values()[data.getInt("stage") % Stage.values().length];
    }

    public static int getPart()
    {
        if(data == null) data = new CompoundNBT();
        return data.getInt("part");
    }

    public static void setStage(int stage)
    {
        data.putInt("stage", stage);
    }

    public static void setStage(int stage, NewEndBiomeProvider provider)
    {
        data.putInt("stage", stage);
        provider.initBiomeLayer();
    }

    public static void setPart(int part)
    {
        data.putInt("part", part);
    }

    public static void addStage()
    {
        setStage(Math.min(getStage() + 1, 3));
    }

    public static void addStage(NewEndBiomeProvider provider)
    {
        setStage(Math.min(getStage() + 1, 3), provider);
    }

    public static void addPart()
    {
        setPart(getPart() + 1);
        if(data.getInt("part") >= 3)
        {
            addStage();
            setPart(0);
        }
    }

    public static void addPart(NewEndBiomeProvider provider)
    {
        setPart(getPart() + 1);
        if(data.getInt("part") >= 3)
        {
            addStage(provider);
            setPart(0);
        }
    }

    public enum Stage
    {
        ASH
            {
                @Override
                public void createTower(CustomEndSpike future, IWorld world, CustomEndSpike.EndSpike spike)
                {
                    if (spike.isGuarded())
                    {
                        BlockPos.Mutable pos = new BlockPos.Mutable();

                        for (int k = -2; k <= 2; ++k)
                        {
                            for (int l = -2; l <= 2; ++l)
                            {
                                for (int i1 = 0; i1 <= 3; ++i1)
                                {
                                    boolean isRight = MathHelper.abs(k) == 2;
                                    boolean ifLeft = MathHelper.abs(l) == 2;
                                    boolean isTop = i1 == 3;
                                    if (isRight || ifLeft || isTop)
                                    {
                                        boolean isNorth = k == -2 || k == 2 || isTop;
                                        boolean flag4 = l == -2 || l == 2 || isTop;
                                        BlockState blockstate = Blocks.IRON_BARS.getDefaultState()
                                                .with(PaneBlock.NORTH, isNorth && l != -2)
                                                .with(PaneBlock.SOUTH, isNorth && l != 2)
                                                .with(PaneBlock.WEST, flag4 && k != -2)
                                                .with(PaneBlock.EAST, flag4 && k != 2);
                                        future.setBlockState(world, pos.setPos(spike.getCenterX() + k, spike.getHeight() + i1, spike.getCenterZ() + l), blockstate);
                                    }
                                }
                            }
                        }
                    }
                }
            }, REDO
            {
                @Override
                public void createTower(CustomEndSpike future, IWorld world, CustomEndSpike.EndSpike spike)
                {
                    BlockPos.Mutable pos = new BlockPos.Mutable();

                    for (int k = -2; k <= 2; ++k)
                    {
                        for (int l = -2; l <= 2; ++l)
                        {
                            for (int i1 = 0; i1 <= 3; ++i1)
                            {
                                boolean isRight = MathHelper.abs(k) == 2;
                                boolean ifLeft = MathHelper.abs(l) == 2;
                                boolean isTop = i1 == 3;
                                if (isRight || ifLeft || isTop)
                                {
                                    BlockState blockstate = PhoenixBlocks.INSTANCE.getARMORED_GLASS().get().getDefaultState();
                                    future.setBlockState(world, pos.setPos(spike.getCenterX() + k, spike.getHeight() + i1, spike.getCenterZ() + l), blockstate);
                                }
                            }
                        }
                    }
                }
            }, REBIRTH
            {
                @Override
                public void createTower(CustomEndSpike future, IWorld world, CustomEndSpike.EndSpike spike)
                {
                    ASH.createTower(future, world, spike);
                }
            },
            AIR
            {
                @Override
                public void createTower(CustomEndSpike future, IWorld world, CustomEndSpike.EndSpike spike)
                {
                    ASH.createTower(future, world, spike);
                }
            };

        public abstract void createTower(CustomEndSpike future, IWorld world, CustomEndSpike.EndSpike spike);
    }
}

File: src/main/kotlin/phoenix/world/StageManager.kt
package phoenix.world

import net.minecraft.block.Blocks
import net.minecraft.block.PaneBlock
import net.minecraft.client.audio.MusicTicker
import net.minecraft.entity.EntityType
import net.minecraft.entity.boss.dragon.EnderDragonEntity
import net.minecraft.nbt.CompoundNBT
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.MathHelper
import net.minecraft.world.IWorld
import net.minecraftforge.common.MinecraftForge
import phoenix.init.PhoenixMusicTracks
import phoenix.init.PhxBlocks.armoredGlass
import phoenix.other.client
import phoenix.other.server
import phoenix.utils.*
import phoenix.world.structures.CustomEndSpike
import kotlin.math.min

object StageManager
{
    private var data: CompoundNBT = CompoundNBT()

    fun read(nbt: CompoundNBT)
    {
        if (nbt.contains("stage_nbt"))
        {
            data = nbt.getCompound("stage_nbt")
        }
        else
        {
            val compound = CompoundNBT()
            compound.putInt("stage", 0)
            compound.putInt("part", 0)
            data = compound
            nbt.put("stage_nbt", compound)
        }
    }

    fun write(compound: CompoundNBT): CompoundNBT
    {
        compound.put("stage_nbt", data)
        return compound
    }

    var stage: Int
        get() = data.getInt("stage")
        set(stage) = data.putInt("stage", stage)
    val stageEnum: Stage
        get() = Stage.values()[data.getInt("stage") % Stage.values().size]
    var part: Int
        get() =  data.getInt("part")
        set(part) = data.putInt("part", part)

    private fun setStage(stage: Int, provider: NewEndBiomeProvider, postEvent : Boolean = true)
    {
        this.stage = stage
        provider.initBiomeLayer()

        if (postEvent)
        {
            client { _, _, world ->
                if (world != null)
                    MinecraftForge.EVENT_BUS.post(ClientChangeStageEvent(world))
            }

            server { server ->
                if (server != null)
                    for (i in server.worlds)
                        MinecraftForge.EVENT_BUS.post(ServerChangeStageEvent(i))
            }
        }
    }

    private fun addStage(provider: NewEndBiomeProvider, postEvent : Boolean = false) = setStage(min(stage + 1, 3), provider, postEvent)

    fun addPart(provider: NewEndBiomeProvider)
    {
        part++
        if (data.getInt("part") >= 3)
        {
            addStage(provider)
            client { _, _, world ->
                if (world != null)
                    MinecraftForge.EVENT_BUS.post(ClientStageUppedEvent(world))
            }

            server { server ->
                if (server != null)
                    for (world in server.worlds)
                        MinecraftForge.EVENT_BUS.post(ServerStageUppedEvent(world))
            }
            part = 0
        }
    }

    enum class Stage
    {
        ASH
        {
            override fun createTower(future: CustomEndSpike, world: IWorld, spike: CustomEndSpike.EndSpike)
            {
                if (spike.isGuarded)
                {
                    val pos = BlockPos.Mutable()
                    for (x in -2..2)
                    {
                        for (z in -2..2)
                        {
                            for (y in 0..3)
                            {
                                val isRight = MathHelper.abs(x) == 2
                                val ifLeft = MathHelper.abs(z) == 2
                                val isTop = y == 3
                                if (isRight || ifLeft || isTop)
                                {
                                    val isNorth = x == -2 || x == 2 || isTop
                                    val flag4 = z == -2 || z == 2 || isTop
                                    val blockstate = Blocks.IRON_BARS.defaultState
                                        .with(PaneBlock.NORTH, isNorth && z != -2)
                                        .with(PaneBlock.SOUTH, isNorth && z != 2)
                                        .with(PaneBlock.WEST, flag4 && x != -2)
                                        .with(PaneBlock.EAST, flag4 && x != 2)
                                    future.setBlockState(
                                        world,
                                        pos.setPos(spike.centerX + x, spike.height + y, spike.centerZ + z),
                                        blockstate
                                    )
                                }
                            }
                        }
                    }
                }
            }

            override val music: MusicTicker.MusicType = MusicTicker.MusicType.END

            override val dragonType: EntityType<out EnderDragonEntity> = EntityType.ENDER_DRAGON
        },
        REDO
        {
            override fun createTower(future: CustomEndSpike, world: IWorld, spike: CustomEndSpike.EndSpike)
            {
                val pos = BlockPos.Mutable()
                for (x in -2..2)
                {
                    for (z in -2..2)
                    {
                        for (y in 0..3)
                        {
                            val isRight = MathHelper.abs(x) == 2
                            val ifLeft = MathHelper.abs(z) == 2
                            val isTop = y == 3
                            if (isRight || ifLeft || isTop)
                            {
                                future.setBlockState(world, pos.setPos(spike.centerX + x, spike.height + y, spike.centerZ + z), armoredGlass.defaultState)
                            }
                        }
                    }
                }
            }
            override val music: MusicTicker.MusicType = PhoenixMusicTracks.REDO_MUSIC
            override val dragonType: EntityType<out EnderDragonEntity> = ASH.dragonType
        },
        REBIRTH
        {
            override fun createTower(future: CustomEndSpike, world: IWorld, spike: CustomEndSpike.EndSpike) = REDO.createTower(future, world, spike)
            override val music: MusicTicker.MusicType = REDO.music
            override val dragonType: EntityType<out EnderDragonEntity> = REDO.dragonType
        },
        AIR
        {
            override fun createTower(future: CustomEndSpike, world: IWorld, spike: CustomEndSpike.EndSpike) = ASH.createTower(future, world, spike)
            override val music: MusicTicker.MusicType = REBIRTH.music
            override val dragonType: EntityType<out EnderDragonEntity> = REBIRTH.dragonType
        };

        abstract fun createTower(future: CustomEndSpike, world: IWorld, spike: CustomEndSpike.EndSpike)

        abstract val music : MusicTicker.MusicType

        abstract val dragonType : EntityType<out EnderDragonEntity>
    }
}
